Ubuntu: reported by Agostino Russo
Upstream: http://lists.gnu.org/archive/html/grub-devel/2009-09/msg00199.html
Description: Handle 64-bit file sizes in NTFS. This should make it possible
 for Wubi to use the loopback module.

diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta2/fs/ntfs.c grub2-1.97~beta2.new/fs/ntfs.c
--- grub2-1.97~beta2/fs/ntfs.c	2009-08-28 17:29:34.000000000 +0100
+++ grub2-1.97~beta2.new/fs/ntfs.c	2009-09-10 19:26:10.000000000 +0100
@@ -543,7 +543,7 @@
       if (!pa[8])
 	mft->size = u32at (pa, 0x10);
       else
-	mft->size = u32at (pa, 0x30);
+	mft->size = u64at (pa, 0x30);
 
       if ((mft->attr.flags & AF_ALST) == 0)
 	mft->attr.attr_end = 0;	/*  Don't jump to attribute list */
diff -Nur -x '*.orig' -x '*~' grub2-1.97~beta2/include/grub/ntfs.h grub2-1.97~beta2.new/include/grub/ntfs.h
--- grub2-1.97~beta2/include/grub/ntfs.h	2008-09-22 05:18:57.000000000 +0100
+++ grub2-1.97~beta2.new/include/grub/ntfs.h	2009-09-10 19:26:10.000000000 +0100
@@ -133,7 +133,7 @@
 {
   struct grub_ntfs_data *data;
   char *buf;
-  grub_uint32_t size;
+  grub_uint64_t size;
   grub_uint32_t ino;
   int inode_read;
   struct grub_ntfs_attr attr;
